### PR TITLE
DEVX-550: Producer/consumer example to Confluent Cloud: .NET/Csharp

### PR DESCRIPTION
**Pull request by mhowlett**
note: there is a bug in 1.0-beta3 which will cause the create topics command to hang, so this example is referencing a ci build in which this is resolved. alternatively this could have been worked around, but the code to do that is a bit ugly. When RC1 is released, we can just update the reference.

### CODE REVIEW

**Comment by ybyzek** [comment]
@mhowlett thanks for the PR!  

Suggestions:
1. Apply this PR to 5.1.1-post (instead of 5.1.2-post)
2. Add an entry and image to the README at https://github.com/confluentinc/examples/tree/5.1.1-post/clients/cloud/README.md

File: clients/cloud/csharp/README.md [hunk 1]
```diff
@@ -0,0 +1,90 @@
+﻿# Overview
+
+Produce messages to and consume messages from [Confluent Cloud](https://www.confluent.io/confluent-cloud/) using the .NET Producer and Consumer.
+
+
+# Prerequisites
+
+* .NET Core 2.1 or higher to run the demo application.
```
Comment (ybyzek) [inline clients/cloud/csharp/README.md]: Please provide link to install instructions (e.g. https://github.com/confluentinc/confluent-kafka-dotnet)

File: clients/cloud/csharp/Program.cs [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+            {
+                case "produce":
+                    await CreateTopicMaybe(topic, 1, 3, config);
+                    Produce(topic, config);
+                    break;
+                case "consume":
+                    Consume(topic, config);
+                    break;
+                default:
+                    Console.WriteLine($"First command line argument must be 'produce' or 'consume'. It was: {mode}");
```
Comment (edenhill) [inline clients/cloud/csharp/Program.cs]: This should be the usage printout.

File: clients/cloud/csharp/Program.cs [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+            consumerConfig.AutoOffsetReset = AutoOffsetReset.Earliest;
+            consumerConfig.EnableAutoCommit = false;
+
+            CancellationTokenSource cts = new CancellationTokenSource();
+            Console.CancelKeyPress += (_, e) => {
+                e.Cancel = true; // prevent the process from terminating.
+                cts.Cancel();
+            };
+
+            using (var consumer = new ConsumerBuilder<string, string>(consumerConfig).Build())
```
Comment (edenhill) [inline clients/cloud/csharp/Program.cs:147]: This is a neat consume loop 👍

File: clients/cloud/csharp/Program.cs [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+                        var cr = consumer.Consume();
+                        totalCount += JObject.Parse(cr.Value).Value<int>("count");
+                        Console.WriteLine($"Consumed record with key {cr.Key} and value {cr.Value}, and updated total count to {totalCount}");
+                    }
+                }
+                catch (OperationCanceledException)
+                {
+                    // Ctrl-C was pressed.
+                    Console.WriteLine("closed");
+                    consumer.Close();
```
Comment (edenhill) [inline clients/cloud/csharp/Program.cs:166]: Id probably want a Close() in a finally block to show how to deal with non-ctrl-c terminations of the loop.

File: clients/cloud/csharp/Program.cs [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+            {
+                var cloudConfig = (await File.ReadAllLinesAsync(configPath))
+                    .Where(line => !line.StartsWith("#"))
+                    .ToDictionary(
+                        line => line.Substring(0, line.IndexOf('=')),
+                        line => line.Substring(line.IndexOf('=') + 1));
+
+                var clientConfig = new ClientConfig
+                {
+                    BootstrapServers = cloudConfig["bootstrap.servers"].Replace("\\", ""),
```
Comment (edenhill) [inline clients/cloud/csharp/Program.cs:55]: Are bootstrap.servers enveloped in quotes? Dont think Ive seen that

File: clients/cloud/csharp/Program.cs [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+                if (certDir != null)
+                {
+                    clientConfig.SslCaLocation = certDir;
+                }
+
+                return clientConfig;
+            }
+            catch
+            {
+                Console.WriteLine($"An error occured reading the config file from: {configPath}");
```
Comment (edenhill) [inline clients/cloud/csharp/Program.cs]: Print the error

File: clients/cloud/csharp/Program.cs [hunk 7]
```diff
@@ -0,0 +1,183 @@
+﻿using Confluent.Kafka;
```
Comment (edenhill) [inline clients/cloud/csharp/Program.cs]: copyright header

File: clients/cloud/csharp/Program.cs [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+
+namespace CCloud
+{
+    class Program
+    {
+        static string ExtractJaasValue(string jaasConfig, string key)
```
Comment (edenhill) [inline clients/cloud/csharp/Program.cs:35]: docstrings what this does (translates the java jaas config for plain auth to .NET client conf properties)

File: clients/cloud/csharp/Program.cs [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+                var cloudConfig = (await File.ReadAllLinesAsync(configPath))
+                    .Where(line => !line.StartsWith("#"))
+                    .ToDictionary(
+                        line => line.Substring(0, line.IndexOf('=')),
+                        line => line.Substring(line.IndexOf('=') + 1));
+
+                var clientConfig = new ClientConfig
+                {
+                    BootstrapServers = cloudConfig["bootstrap.servers"].Replace("\\", ""),
+                    BrokerVersionFallback = "0.10.0.0",
```
Comment (edenhill) [inline clients/cloud/csharp/Program.cs]: This and FallbackMs are not required with librdkafka v1.0.0

**Review by edenhill — CHANGES_REQUESTED** [review]

File: clients/cloud/csharp/Program.cs (hunk 2 above)
Comment (mhowlett) [reply clients/cloud/csharp/Program.cs]: yep, i did think that but was lazy and mimicked the other examples. changed.

File: clients/cloud/csharp/Program.cs (hunk 4 above)
Comment (mhowlett) [reply clients/cloud/csharp/Program.cs:166]: big whoops! 1. great point - `Close` is in the wrong spot 2. that 'closed' printf is not supposed to be there. 3. I forgot to even pass the cancellation token into the `Consume` method.

File: clients/cloud/csharp/Program.cs (hunk 5 above)
Comment (mhowlett) [reply clients/cloud/csharp/Program.cs:55]: `:` is escaped in the ccloud config, this is accounting for that. better treatment of escaped characters would be safer obviously, but i thought v. minor risk of any issues, and not worth the effort.

File: clients/cloud/csharp/Program.cs (hunk 8 above)
Comment (mhowlett) [reply clients/cloud/csharp/Program.cs:35]: fair enough

File: clients/cloud/csharp/Program.cs (hunk 9 above)
Comment (mhowlett) [reply clients/cloud/csharp/Program.cs]: awesome

File: clients/cloud/csharp/Program.cs [hunk 10]
```diff
@@ -136,26 +152,31 @@ static void Consume(string topic, ClientConfig config)
...
+                finally
+                {
                     consumer.Close();
                 }
             }
         }
 
+        static void PrintUsage()
+        {
+            Console.WriteLine("usage: .. <produce|consume> topic configPath [certDir]");
```
Comment (edenhill) [inline clients/cloud/csharp/Program.cs]: So, the typical command syntax is `<..>` for variable fields, thus:
`Usage: .. produce|consume <topic> <configPath> [<certDir>]`

**Review by edenhill — APPROVED** [review]
LGTM.

Tiny nit on syntaxing

File: clients/cloud/csharp/Program.cs (hunk 10 above)
Comment (mhowlett) [reply clients/cloud/csharp/Program.cs]: thanks for the education - didn't know that, wondered... (will change)

**Comment by mhowlett** [comment]
@ybyzek - have finished updating this - feel free to merge if you have no more suggestions.

**Pull request merged**